scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/StepSimulation.java

```diff
@@ -0,0 +1,255 @@
+      var activity = act.getValue().getLeft();
+      var name = act.getValue().getRight();
+      if(start.shorterThan(startTime)) {
+        this.simulateAct(activity, startTime, name);
+      }
+    }
+  }
+
+  private void insertAllActsAfter(LinkedHashMap<Duration, Pair<SerializedActivity, String>> acts, Duration startTime){
+    for(var act : acts.entrySet()){
+      var start = act.getKey();
+      var activity = act.getValue().getLeft();
+      var name = act.getValue().getRight();
+      if(start.longerThan(startTime) || start.isEqualTo(startTime)) {
+        this.simulateAct(activity, startTime, name);
+      }
+    }
+  }
+
+  //WARNING computes the results for the whole horizon until current time them from beginning
+  public SimulationResults getSimulationResults(){
+    if(areLastSimResultsDirty || lastSimResults == null) {
+      lastSimResults = engine.computeResults(
+          engine,
+          Instant.now(),
+          curTime,
+          new HashMap<String, ActivityInstanceId>(),
+          timeline,
+          missionModel);
+    }
+    return lastSimResults;
+  }
+
+  //WARNING computes the results for the whole horizon until current time them from beginning
+  public SimulationResults getSimulationResultsUntil(Duration endTime){
+    if(areLastSimResultsDirty || lastSimResults == null) {
+      lastSimResults = engine.computeResults(
+          engine,
+          Instant.now(),
+          endTime,
+          new HashMap<String, ActivityInstanceId>(),
+          timeline,
+          missionModel);
+    }
+    return lastSimResults;
+  }
+
+
+  private void simulateAct(SerializedActivity activity, Duration startTime, String name){
+
+    final var schedule = Map.of(name, Pair.of(startTime, activity));
+
+    final var controlTask = new ControlTask(schedule, curTime);
+    // Schedule the control task.
+    final var control = engine.initiateTask(curTime, controlTask);
+    actToId.put(activity, control);
+
+    engine.scheduleTask(control, curTime);
+
+    while (!engine.isTaskComplete(control)) {
+      final var batch = engine.extractNextJobs(Duration.MAX_VALUE);
+
+      // Increment real time, if necessary.
+      final var delta = batch.offsetFromStart().minus(curTime);
+      curTime = batch.offsetFromStart();
+      timeline.add(delta);
+      // TODO: Advance a dense time counter so that future tasks are strictly ordered relative to these,
+      //   even if they occur at the same real time.
+
+      // Run the jobs in this batch.
+      final var commit = engine.performJobs(batch.jobs(), cells, curTime, Duration.MAX_VALUE, missionModel);
+      timeline.add(commit);
+    }
+    areLastSimResultsDirty = true;
+    activitiesInserted.put(startTime, Pair.of(activity, name));
+  }
+
+  public Duration getActivityDuration(SerializedActivity act){
+    //peek into engine
+    return engine.getTaskDuration(actToId.get(act));
+  }
+
+  public boolean activityHasFinished(SerializedActivity act){
```

Comment:
Is this method used anywhere, or are there any production use-cases expected for it?

Reply:
No it can be removed for now.